### PR TITLE
Train and eval data splits (Nerfbusters)

### PR DESCRIPTION
**Pull request by FrederikWarburg**
tl;dr. Add supports of datasets with various train/eval splits + visualization of eval image in viewer + enable ns-process-data to take two videos, one used for training and the other for evaluation.

Added a new flag:
--eval-mode "train-split-fraction", "eval-frame-index", "eval-interval"

that basically decides if the eval set is a fraction of the images (set to default), if the eval/train split is based on the names of the files, or every nth frame is used for evaluation.

ns-train nerfacto --data ../nerfbusters/data/nerfbusters-dataset-renamed/aloe/ nerfstudio-data --eval-mode "train-split-fraction"

Also, added visualizations of the eval frames. See screenshots for the 3 settings.

Screenshot 2023-07-12 at 3 41 36 PM
Screenshot 2023-07-12 at 3 39 57 PM
Screenshot 2023-07-12 at 3 38 26 PM

For ns-process-data, I've tested by running:

ns-process-data video --data data/aloe/aloe.mp4 --output-dir data/processed-aloe-single --no-gpu

and

ns-process-data video --data data/aloe/aloe.mp4 data/aloe/aloe-eval.mp4 --output-dir data/processed-aloe

which generates the expected outputs, i.e. when only a single path is given, it is the same as normally, but if two paths are provided, the images will be saved in the same folder, with prefix either frame_train_ or frame_eval_.

### CODE REVIEW

File: nerfstudio/data/dataparsers/nerfstudio_dataparser.py [hunk 1]
```diff
@@ -58,8 +55,12 @@ class NerfstudioDataParserConfig(DataParserConfig):
     """The method to use to center the poses."""
     auto_scale_poses: bool = True
     """Whether to automatically scale the poses to fit in +/- 1 bounding box."""
+    eval_mode: Literal["train-split-fraction", "eval-frame-index", "eval-interval"] = "train-split-fraction"
+    """The method to use for splitting the dataset into train and eval."""
     train_split_fraction: float = 0.9
-    """The fraction of images to use for training. The remaining images are for eval."""
+    """The index of the frames to use for eval. Eval frames are names frame_1_00001.png, frame_1_00002.png, etc."""
```
Comment (ethanweber) [inline nerfstudio/data/dataparsers/nerfstudio_dataparser.py:61]: Is this outdated if you are using train and eval strings?

File: nerfstudio/process_data/images_to_nerfstudio_dataset.py [hunk 2]
```diff
@@ -47,20 +45,24 @@ def main(self) -> None:
...
-            self.data = equirect_utils.generate_planar_projections_from_equirectangular(
-                self.data, pers_size, self.images_per_equirect, crop_factor=self.crop_factor
+            self.data = (
+                equirect_utils.generate_planar_projections_from_equirectangular(
+                    self.data[0], pers_size, self.images_per_equirect, crop_factor=self.crop_factor
+                ),
             )
             self.camera_type = "perspective"
 
+        assert len(self.data) == 1, "Only one data source is supported for images"
```
Comment (ethanweber) [inline nerfstudio/process_data/images_to_nerfstudio_dataset.py:57]: Should we add support for this in this PR too?

File: nerfstudio/process_data/video_to_nerfstudio_dataset.py [hunk 3]
```diff
@@ -46,25 +44,49 @@ def main(self) -> None:
...
-                self.data,
+                self.data[0],
                 image_dir=self.image_dir,
                 num_frames_target=self.num_frames_target,
                 crop_factor=self.crop_factor,
                 verbose=self.verbose,
+                image_prefix="frame_train_" if len(self.data) == 2 else "frame_",
             )
 
+            if len(self.data) == 2:
```
Comment (ethanweber) [inline nerfstudio/process_data/video_to_nerfstudio_dataset.py:68]: We should probably add more comments to explain this?

File: nerfstudio/scripts/process_data.py [hunk 4]
```diff
@@ -69,7 +65,7 @@ def main(self) -> None:
 
         summary_log = []
 
-        record3d_image_dir = self.data / "rgb"
+        record3d_image_dir = self.data[0] / "rgb"
```
Comment (ethanweber) [inline nerfstudio/scripts/process_data.py:68]: This feels weirdly necessary because of how we are typing data. Maybe there is a way around this?

**Review by ethanweber — COMMENTED** [review]
Leaving some comments.

File: nerfstudio/process_data/images_to_nerfstudio_dataset.py (hunk 2 above)
Comment (FrederikWarburg) [reply nerfstudio/process_data/images_to_nerfstudio_dataset.py:57]: Maybe we can split into a seperate PR?

File: nerfstudio/scripts/process_data.py (hunk 4 above)
Comment (FrederikWarburg) [reply nerfstudio/scripts/process_data.py:68]: Yea, it is slightly annoying... Instead of having data be a Tuple of train and eval paths, we could keep data as is, and create a new argument eval_data. Let me try this.

File: nerfstudio/process_data/base_converter_to_nerfstudio_dataset.py [hunk 5]
```diff
@@ -19,16 +19,19 @@
...
 @dataclass
 class BaseConverterToNerfstudioDataset(ABC):
-    """Base class to process images or video into a nerfstudio dataset"""
+    """Base class to process images or video into a nerfstudio dataset."""
 
     data: Path
     """Path the data, either a video file or a directory of images."""
     output_dir: Path
     """Path to the output directory."""
+    eval_data: Union[Path, None] = None
```
Comment (FrederikWarburg) [inline nerfstudio/process_data/base_converter_to_nerfstudio_dataset.py:33]: Optional[Path]

File: nerfstudio/process_data/images_to_nerfstudio_dataset.py [hunk 6]
```diff
@@ -54,6 +52,8 @@ def main(self) -> None:
             )
             self.camera_type = "perspective"
 
+        assert self.eval_data is None, "Only one data source is supported for images"
```
[the remark this answers is not in the document]
Comment (FrederikWarburg) [reply nerfstudio/process_data/images_to_nerfstudio_dataset.py:55]: Implement for images also.

File: nerfstudio/scripts/process_data.py [hunk 7]
```diff
@@ -258,6 +254,8 @@ def main(self) -> None:
         image_dir = self.output_dir / "images"
         image_dir.mkdir(parents=True, exist_ok=True)
 
+        assert self.eval_data is None, "Only one dataset can be processed at a time"
```
Comment (FrederikWarburg) [inline nerfstudio/scripts/process_data.py:257]: delete

File: nerfstudio/data/dataparsers/nerfstudio_dataparser.py [hunk 8]
```diff
@@ -58,8 +55,12 @@ class NerfstudioDataParserConfig(DataParserConfig):
     """The method to use to center the poses."""
     auto_scale_poses: bool = True
     """Whether to automatically scale the poses to fit in +/- 1 bounding box."""
+    eval_mode: Literal["train-split-fraction", "eval-frame-index", "eval-interval"] = "train-split-fraction"
```
Comment (FrederikWarburg) [inline nerfstudio/data/dataparsers/nerfstudio_dataparser.py:58]: eval_mode: fraction, filename, interval

Add a description in docs string

File: nerfstudio/data/dataparsers/nerfstudio_dataparser.py [hunk 9]
```diff
@@ -105,9 +106,19 @@ def _generate_dataparser_outputs(self, split="train"):
         width = []
         distort = []
 
+        # sort the frames by fname
+        # they do this in mipnerf360 code
```
Comment (FrederikWarburg) [inline nerfstudio/data/dataparsers/nerfstudio_dataparser.py:110]: drop

**Review by ethanweber — APPROVED** [review]
LGTM @FrederikWarburg. I made changes such that it supports images and videos but not equirectangular types. I had to modify code here and there to avoid breaking some recent changes to main.

**Comment by ethanweber** [comment]
Here are some final change updates. Processing the data works for both `ns-process-data images` or `ns-process-data video`, as mentioned in the documentation.

```python
eval_mode: Literal["fraction", "filename", "interval", "all"] = "fraction"
"""
The method to use for splitting the dataset into train and eval. 
Fraction splits based on a percentage for train and the remaining for eval.
Filename splits based on filenames containing train/eval.
Interval uses every nth frame for eval.
All uses all the images for any split.
"""
```

In the viewer, it will show the eval images with a red border.

**Comment by RoyAmoyal** [comment]
@FrederikWarburg 

My data:
Images_train_eval/ 

names of files:
MTN_5896_train.png
MTN_5897_eval.png 
etc

but I am still getting
raise ValueError("frame should contain train/eval in its name to use this eval-frame-index eval mode")
